---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# PhenoGenRLib

An easy way to study associations between single-nucleotide polymorphisms and phenotypic traits given a dataset of variants.

<!-- badges: start -->
<!-- badges: end -->

---

The goal of PhenoGenRLib is to simplify nucleotide variant analysis. As next generational sequencing (NGS) begins taking off, more and more data is readily available to be used. As a direct product of the growing popularity of the new machine, there is an increasing amount of data that has yet been used to it's fullest potential. 

PhenoGenRLib promises to provide simple ways of loading [variant call format (VCF)](https://en.wikipedia.org/wiki/Variant_Call_Format) files, a commonly used bioinformatics file format capable of storing all forms of variants given a sequence. The format is also quite extensible capable of generalized metadata storage. linking them with sample metadata, and lastly, running associative studies by applying the metadata. Furthermore, PhenoGenRLib contains simple functions for rapidly resolving for rsIDs based of the human chromosome and position and fetching from the National Center for Biotechnology Information (NCBI) Entrez portal.

The package is developed under the following environment:

- R version 4.3.2 (2023-10-31)
- Platform: x86_64-conda-linux-gnu (64-bit)
- Running under: Ubuntu 22.04.3 LTS

## Installation

You can install the development version of PhenoGenRLib like so:

``` r
install.packages("devtools")
require("devtools")
devtools::install github("RealYHD/PhenoGenRLib", build vignettes = TRUE)
library("PhenoGenRLib")
```

To run the shiny app:

``` r
runPhenoGenRLib()
```

## Overview

<!-- TODO -->


# Contributions

PhenoGenRLib stands on the shoulder of giants, and it would be a disservice to not name them:

- Thank you Syed Haider et al. for providing bedr. It was greatly helpful in simplifying the data ingress features.
- Thanks to the entire Biomart Team for providing an awesome and easy to use interface to large public databases!
- ggplot2 was very helpful in generating figures. Thanks to Wickham et. al!
- This entire project wouldn't have been possible without the help of the TidyVerse team. Despite not using every single package from their library, much work and diagnostic made use of their tools.
- Tibble helped simplify data storage and accession. Thanks Muller et. al!

No generative AI was used for this project directly, however, learning about how R works and how some of the syntax differs from other languages was aided by ChatGPT.

This was a BCB410H1 UofT Bioinformatics project by Harrison Deng. 

# Citations

```
MÃ¼ller K, Wickham H (2023). _tibble: Simple Data Frames_. R package
  version 3.2.1, <https://CRAN.R-project.org/package=tibble>.

Haider S, Waggott D, C. Boutros P (2019). _bedr: Genomic Region
  Processing using Tools Such as 'BEDTools', 'BEDOPS' and 'Tabix'_. R
  package version 1.0.7, <https://CRAN.R-project.org/package=bedr>.

BioMart and Bioconductor: a powerful link between biological
  databases and microarray data analysis. Steffen Durinck, Yves Moreau,
  Arek Kasprzyk, Sean Davis, Bart De Moor, Alvis Brazma and Wolfgang
  Huber, Bioinformatics 21, 3439-3440 (2005).

H. Wickham. ggplot2: Elegant Graphics for Data Analysis.
  Springer-Verlag New York, 2016.

Wickham H, Hester J, Bryan J (2023). _readr: Read Rectangular Text
  Data_. R package version 2.1.4,
  <https://CRAN.R-project.org/package=readr>.

```

# Acknowledgements

This package was developed as part of an assessment for 2023 BCB410H: Applied Bioinformatics course at the University of Toronto, Toronto, CANADA. PhenoGenRLib welcomes issues, enhancement requests, and other contributions. To submit an issue, use the GitHub issues.



## Getting Started

To get started, have a datasheet ready in the form of a `CSV`. This datasheet should at the very least, contain one column, where each row in that column contains the filename of the VCF including the `.vcf`. For the following example, we will assume that such a file is called `huntingtons_datasheet_shortened.csv` and is located at `./inst/extdata/huntingtons_datasheet_shortened.csv` with the column containing the VCF filenames being named `vcfs`. We will also need the location of the VCFs. Let's assume they can be found at the same place as the metadata CSV `./inst/extdata/`. Then:

```{r example}
library(PhenoGenRLib)
variants <- PhenoGenRLib::linkVariantsWithMetadata(
  metadata = "inst/extdata/huntingtons_datasheet_shortened.csv",
  vcfDir = "inst/extdata/",
  vcfColName = "vcfs"
)
```

Checkout the documents and vignettes for where to go from here!
