---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# PhenoGenRLib

<!-- badges: start -->
<!-- badges: end -->

The goal of PhenoGenRLib is to simplify nucleotide variant analysis.

As next generational sequencing (NGS) begins taking off, more and more data is readily available to be used. Arguably, there is an overabundance of data that has yet been used to it's fullest potential. PhenoGenRLib promises to provide simple ways of loading VCFs, associating them with sample metadata, and lastly, running associative studies by applying the metadata.

## Installation

You can install the development version of PhenoGenRLib like so:

``` r
require("devtools")
devtools::install github("RealYHD/PhenoGenRLib",
build vignettes = TRUE)
library("PhenoGenRLib")
```

## Getting Started

To get started, have a datasheet ready in the form of a `CSV`. This datasheet should at the very least, contain one column, where each row in that column contains the filename of the VCF including the `.vcf`. For the following example, we will assume that such a file is called `huntingtons_datasheet_shortened.csv` and is located at `./inst/extdata/huntingtons_datasheet_shortened.csv` with the column containing the VCF filenames being named `vcfs`. We will also need the location of the VCFs. Let's assume they can be found at the same place as the metadata CSV `./inst/extdata/`. Then:

```{r example}
library(PhenoGenRLib)
PhenoGenRLib::linkVariantsWithMetadata(
  metadata = "inst/extdata/huntingtons_datasheet_shortened.csv",
  vcfDir = "inst/extdata/",
  vcfColName = "vcfs"
)
```

Checkout the documents and vignettes for where to go from here!